keps/sig-network/2030-endpointslice-subsetting/README.md

```diff
@@ -0,0 +1,212 @@
+locations.
+
+As clusters and services grow, we’re also seeing that the current proxy
+implementations are stretched to their limits as all endpoints for the entire
+cluster are tracked independently by each node.
+
+The best user experience seems like it would be to have the platform - with a
+first class controller and/or a provider specific implementation - intelligently
+prioritize endpoints based on topology, capacity, and any other useful cost
+metrics to aid in traffic shaping, or just to reduce the amount of global
+resource tracking required by every node. However, we are currently missing a
+way to allow endpoints to be targeted to specific subsets of nodes.
+
+### Goals
+
+- Provide the building blocks to allow EndpointSlices to target specific subsets
+  of nodes.
+- EndpointSlices subsetting will be fully backwards compatible for older
+  consumers of the EndpointSlice API.
+- Design is flexible enough for multiple implementations and experimentation.
+- Minimal duplication of data.
+- Room for future enhancements, for example weighted endpoints or Slices.
+
+### Non-Goals
+
+- Define how subsetting should be used.
+- Design the controller responsible for subsetting endpoints.
+- An API for telling controllers how a service should be subsetted.
+
+Many of these are being tackled by the follow up [KEP
+#2004](https://github.com/kubernetes/enhancements/issues/2004).
+
+## Proposal
+
+Two new topology based labels will be introduced for EndpointSlices to support
+subsetting:
+
+```
+endpointslice.kubernetes.io/for-zone
+endpointslice.kubernetes.io/for-region
+```
+
+In the future this pattern may be expanded to include other concepts or
+topologies. A simple pattern like this will allow EndpointSlices to be delivered
+to consumers in a specific zone or region.
+
+
+### Risks and Mitigations
+
+This approach does not allow a single EndpointSlice to target multiple zones or
+regions. Any approach that enabled that would be significantly more complicated.
+The initial proposal in [KEP
+#2004](https://github.com/kubernetes/enhancements/issues/2004) suggests that
+this won't be an issue. 
+
+## Design Details
+
+### API Consumption
+
+EndpointSlice API consumers can use the following selectors to consume
+EndpointSlices:
+
+#### Select EndpointSlices Without Labels for Backwards Compatibility
+```
+matchExpressions:
+  - {key: endpointslice.kubernetes.io/for-zone, operator: DoesNotExist}
+  - {key: endpointslice.kubernetes.io/for-region, operator: DoesNotExist}
+```
+
+#### Select EndpointSlices With Matching Zone
+```
+matchLabels:
+  endpointslice.kubernetes.io/for-zone: example-zone
+```
+
+#### Select EndpointSlices With Matching Region
+```
+matchLabels:
+  endpointslice.kubernetes.io/for-region: example-region
+```
+
+#### Kube-Proxy
+When the `EndpointSliceSubsetting` feature gate is set to true on `Kube-Proxy`,
+it will use these selectors to filter EndpointSlices.
```

Comment:
Will it have two separate watches (for-zone and for-region)?

Reply:
Provided more detail below, open to whatever approach you think would be best here.

Reply:
Sorry - I don't see that anywhere - can you point me to the text you added?

Reply:
That was confusing, I was just meaning that my answer to your comment below also applied here. As proposed, this would require 2-3 watchers per kube-proxy (depending on if we add a watch for region).

Reply:
I can't think of any alternatives. Would be awesome if there was an effective way to do an `OR` with a k8s selector. @wojtek-t Do you think the additional watchers added by this proposal would be preferable to the single broader watch we currently have?

Reply:
You could get this down to two watches: one for untagged, and one for a combination of zone + region. Then, if an EPSlice is for a specific region, then you tag it with a synthetic key of zone+region, for all zones in the region.

Reply:
I've been trying to find a way to make the multi-zone delivery work here but I can't think of a clean way to do that. What would this approach look like? I think it's important to be able to deliver to an entire region without a full concept of all zones contained within that region.